include/cuco/detail/bloom_filter/bloom_filter.inl

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/storage/storage_base.cuh>
+#include <cuco/utility/cuda_thread_scope.cuh>
+
+#include <cuda/atomic>
+#include <cuda/stream_ref>
+
+#include <cstdint>
+
+namespace cuco {
+
+template <class Key,
+          class Block,
+          class Extent,
+          cuda::thread_scope Scope,
+          class Hash,
+          class Allocator>
+__host__ bloom_filter<Key, Block, Extent, Scope, Hash, Allocator>::bloom_filter(
+  Extent num_blocks,
+  uint32_t pattern_bits,
```

Comment:
It might be worth considering putting `pattern_bits` in the template parameters (or embedded in there as a static constant in one of the template parameters, etc.), in order to prevent accidental use of two BFs with different `pattern_bits` in an operation like union or intersection at compile time.

Reply:
We could use an `std::extent`-like approach and make the parameter static as opt-in like so: `pattern_bits<8>{}` for a static parameter and `pattern_bits<DYNAMIC>{8}` for a dynamic parameter. In the former case the parameter would be part of the filter type and would throw a substitution error in case we try to assign a dynamic `pattern_bits`.

Reply:
I think the policy concept addresses this from a different angle: If we make the `pattern_bits` parameter a `static constexpr` attribute of the specific policy, then it's already part of the type. Thus, I will resolve this thread.